[AJ-585, AJ-590, AJ-591, AJ-592, AJ-595] WDS-powered data tables

davidangb opened the pull request:
TL;DR:
* I would like to merge this code to prevent it from being a long-running - and potentially stale - feature branch, even though it still looks for a WDS running on localhost.
* The new code is entirely feature-flagged. If the flag is not enabled, no new code will execute and users will see no difference.

---

This follows up on #3448. That other PR refactored parts of the Data tab and the DataTable component to not hardcode ajax calls directly to Rawls. This PR modifies the Data tab/Data Table to call either Rawls or WDS, which is necessary to support Azure moving forward.

_If_ you have the `workspace-data-service` feature flag enabled, and _if_ there is a WDS running on `http://localhost:8080`, then display data tables powered by WDS.

Else, if the feature flag is enabled but WDS cannot be reached, display an error message that WDS is unavailable.

WDS is _not_ currently reachable in workspaces the dev environment; we are working on making that available, but this is a significant effort. Until then, the only way to make this code work is to run a local instance of WDS on `http://localhost:8080` and have your UI contact it there.

Once WDS is working in the dev env, this code will need updates. It will also need updates to handle higher environments. Until then, do not enable the feature flag in non-dev/non-local environments.

![Screenshot 10-4-2022 at 09 40 AM](https://user-images.githubusercontent.com/6041577/193835921-8c3af03a-0ea2-4611-8b7d-0739d81c262c.png)

TODOs for future PRs:
- [ ] disable row selection
- [ ] decide what label to use for the `sys_name` column
- [ ] disable Rawls-based data tables if WDS is enabled
- [ ] enable the Data tab for Azure workspaces
- [ ] follow up on TODOs about reloading and MixPanel

davidangb committed:
detect running WDS on localhost

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
cleaner conditional for contacting WDS

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
display record types and counts in left nav

davidangb committed:
nonfunctional: progress on data table

davidangb committed:
working data table?

davidangb committed:
pagination for WDS

davidangb committed:
isolate the diff between WDS and Entity Service ajax calls

davidangb committed:
remove extraneous changes

davidangb committed:
disable search fields

davidangb committed:
whoops

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
WIP support for delete-table

davidangb committed:
provider fixes

davidangb committed:
EntityServiceDataProvider

davidangb committed:
start of data providers in Data.js

davidangb committed:
delete-table now uses provider

davidangb committed:
some code cleanup

davidangb committed:
more code cleanup

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
fixes from merge

davidangb committed:
WIP ajax TSV download

davidangb committed:
TSV download

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
disable ajax-download for Entity Service

davidangb committed:
factor out transform methods in WDSDataProvider

davidangb committed:
rename WDSDataProvider -> WDSDataTableProvider

davidangb committed:
inline definition for recordattributes and metadata

davidangb committed:
unit tests for WDSDataTableProvider

davidangb committed:
whoops

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
remove unused import

davidangb committed:
EntityQueryOptions props not optional

davidangb commented on src/libs/ajax.js:
```diff
@@ -892,7 +892,7 @@ const Workspaces = signal => ({
       getEntities: entityType => fetchRawls(`${root}/entities/${entityType}`, _.merge(authOpts(), { signal })).then(r => r.json()),
 
-      getEntitiesTsv: entityType => fetchOrchestration(`api/workspaces/${namespace}/${name}/entities/${entityType}/tsv?model=flexible`, _.mergeAll([authOpts(), { signal }])).then(r => r.text()),
+      getEntitiesTsv: entityType => fetchOrchestration(`api/workspaces/${namespace}/${name}/entities/${entityType}/tsv?model=flexible`, _.mergeAll([authOpts(), { signal }])),
```
before this PR, `getEntitiesTsv` was unused. As of this PR, I'm using it, but  the `.then(r => r.text())` is wrong and should be removed

nawatts replied:
I think I added this for data table versions. Nit: some Ajax methods return a [Response](https://developer.mozilla.org/en-US/docs/Web/API/Response) object and some return the response content (usually via Response.json). I think it would be nice if we were more consistent about that. It looks like you're using the blob form of the response content. Could this function return that blob instead of the Response object? Or could this return text and the consumer that needs a Blob construct one from that text?

davidangb replied:
I've updated so this method returns the blob.

davidangb committed:
cleanup

davidangb commented on src/components/data/WDSContent.js:
```diff
@@ -0,0 +1,55 @@
+  // State
+  const [refreshKey] = useState(0)
+
+  // TODO: AJ-655 something in state management is off; reloading the page while a WDS table is displayed causes errors
```
will tackle this in a future PR

davidangb commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -321,13 +324,19 @@ const DataTableActions = ({ workspace, tableName, rowCount, entityMetadata, onRe
+              // TODO: this overrides the filename specified by the WDS API. Is that ok?
+              dataProvider.downloadTsv(signal, tableName).then(blob => FileSaver.saveAs(blob, `${tableName}.tsv`))
+            }
+            // TODO: AJ-656 add Entity Service vs. WDS indicator to mixpanel event
```
will tackle this in a future PR

davidangb commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -418,6 +427,7 @@ const DataTableActions = ({ workspace, tableName, rowCount, entityMetadata, onRe
       onConfirm: Utils.withBusyState(setLoading)(async () => {
         try {
           await dataProvider.deleteTable(tableName)
+          // TODO: AJ-656 add WDS vs. Entity Service property to the mixpanel event
```
will tackle this in a future PR

davidangb commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -321,13 +324,19 @@ const DataTableActions = ({ workspace, tableName, rowCount, entityMetadata, onRe
+              downloadForm.current.submit()
+            } else if (dataProvider.features.supportsTsvAjaxDownload) {
+              // TODO: this overrides the filename specified by the WDS API. Is that ok?
+              dataProvider.downloadTsv(signal, tableName).then(blob => FileSaver.saveAs(blob, `${tableName}.tsv`))
```
we override filenames elsewhere in TSV download paths, so I'm assuming this is ok

davidangb commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -321,13 +324,19 @@ const DataTableActions = ({ workspace, tableName, rowCount, entityMetadata, onRe
+            } else if (dataProvider.features.supportsTsvAjaxDownload) {
+              // TODO: this overrides the filename specified by the WDS API. Is that ok?
+              dataProvider.downloadTsv(signal, tableName).then(blob => FileSaver.saveAs(blob, `${tableName}.tsv`))
+            }
```
two comments on the TSV download:
1) we override filenames elsewhere in TSV download paths, so I'm assuming this is ok
2) we're using the `FileSaver` approach for WDS download because it prevents the need to support a cookie-authed or form-post-authed backend API. The `FileSaver` approach DOES work for Entity Service, but I am hesitant to make the change to something that is already working well.

nawatts replied:
If we're going to use the Ajax download version for WDS, I'd vote for using it for both cases and scrapping the cookie authed endpoints, but that could certainly be done later.

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
lint fixes

davidangb committed:
move tsv blobification into ajax.js

calypsomatic commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,118 @@
+  id: string,
+  type: string,
+  attributes: { [index: string]: any }, // truly "any" here; the backend Java representation is Map<String, Object>
+  metadata?: { [index: string]: any } // truly "any" here; the backend Java representation is Map<String, Object>
```
This should be taken out now that we've removed it from the backend

calypsomatic commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -732,14 +759,44 @@ const WorkspaceData = _.flow(
+                        onDeleteTable: tableName => {
+                          setSelectedData(undefined)
+                          setWdsSchema(_.remove(typeDef => typeDef.name === tableName, wdsSchema))
+                          forceRefresh() // TODO: this probably isn't correct
```
Why isn't this correct?

davidangb replied:
ah good catch. It's related to https://github.com/DataBiosphere/terra-ui/pull/3429/files#diff-8c26354dc5cab7f1d3c38c6ea2ae1a195a5bcfff6668b119a7d7de3a7ef77ca5R21; I have updated the code comment to reflect that.

davidangb committed:
remove metadata from WDS response

davidangb committed:
update code comment

nawatts commented on src/libs/ajax.js:
```diff
@@ -1226,6 +1226,27 @@ const Surveys = signal => ({
   })
 })
 
+const WorkspaceDataService = signal => ({
```
Can we split this out into a separate file following the pattern of ajax/Apps.js, ajax/AzureStorage.js, etc?

nawatts replied:
Separating it would also allow writing it in TypeScript and having types for these functions would remove some of the need to declare types where these functions are called in WDSDataTableProvider.

msilva-broad replied:
yes, definitely have this be broken out to /libs/ajax/WorkspaceDataService.js or better yet /libs/ajax/WorkspaceDataService.ts.  I also second Nick's recommendation to 'promote' some of the data related types elsewhere to this .ts file.

nit:  Should the name be WorkspaceData to align with the other namings?

msilva-broad replied:
also, once broken out, the methods need unit tests (with mocked fetch to avoid real network calls).  Having the broken out file be .ts will help with type safety on the unit tests :)

nawatts commented on src/libs/ajax.js:
```diff
@@ -1226,6 +1226,27 @@ const Surveys = signal => ({
 })
 
+const WorkspaceDataService = signal => ({
+  getSchema: async instanceId => {
+    const res = await fetchWDS(`${instanceId}/types/v0.2`, _.merge(authOpts(), { signal }))
+    return res.json()
+  },
```
Alternatively, these can be written more concisely like:

```suggestion
  getSchema: instanceId => fetchWDS(`${instanceId}/types/v0.2`, _.merge(authOpts(), { signal })).then(r => r.json()),
```

msilva-broad replied:
probably a preference thing... I prefer the async await style here.... and I like having the const on it's own line that is easier to set a breakpoint on before the return.... (in the event of a local or prod issue that needs debugging)

nawatts replied:
Agreed it's a preference thing, I won't request changes for one way or the other.

nawatts commented on src/libs/ajax/data-table-providers/EntityServiceDataTableProvider.ts:
```diff
@@ -40,4 +41,8 @@ export class EntityServiceDataTableProvider implements DataTableProvider {
     return await Ajax().Workspaces.workspace(this.namespace, this.name).deleteEntitiesOfType(entityType)
   }
+
+  downloadTsv: DownloadTsvFn = async (signal: AbortSignal, entityType: string) => {
+    return await Ajax(signal).Workspaces.workspace(this.namespace, this.name).getEntitiesTsv(entityType)
+  }
```
This function has parameter types specified two ways (once via DownloadTsvFn and once via AbortSignal/string). Alternatively, it could be written as one of:

```suggestion
  downloadTsv: DownloadTsvFn = async (signal, entityType) => {
    return await Ajax(signal).Workspaces.workspace(this.namespace, this.name).getEntitiesTsv(entityType)
  }
```

```suggestion
  downloadTsv = async (signal: AbortSignal, entityType: string): Promise<Blob> => {
    return await Ajax(signal).Workspaces.workspace(this.namespace, this.name).getEntitiesTsv(entityType)
  }
```

I like the second since the types are more readily apparent.

nawatts replied:
Also could be written more concisely as:

```suggestion
  downloadTsv: DownloadTsvFn = (signal: AbortSignal, entityType: string) => Ajax(signal).Workspaces.workspace(this.namespace, this.name).getEntitiesTsv(entityType)
```

msilva-broad replied:
my vote is for nick's 2nd suggestion here... the method signature is already enforced by the DataTableProvider interface marked as implemented by the class here.  So the nicest ergonomics is to have the function args and return types inline as in nick's 2nd suggestion.  This is the style I default to myself since many es-linters will complain if those type declarations are missing. And it's a nice immediate form of documentation.

nawatts commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+  })
+})
+
+export {}
```
Looks like this line could be removed.

nawatts commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,117 @@
+interface RecordResponse {
+  id: string,
+  type: string,
+  attributes: { [index: string]: any } // truly "any" here; the backend Java representation is Map<String, Object>
```
I wonder if `unknown` would be preferable here, so that TS informs you that you need to figure out what type the attribute is before using it.

https://www.typescriptlang.org/docs/handbook/2/functions.html#unknown

msilva-broad replied:
I'd also consider using the very convenient Record<Key, Value> type.... so:

`  attributes: Record<string, unknown>`

nawatts commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,117 @@
+    supportsFiltering: false
+  }
+
+  transformPage: (arg0: RecordQueryResponse, arg1: string, arg2: EntityQueryOptions) => EntityQueryResponse = (wdsPage: RecordQueryResponse, recordType: string, queryOptions: EntityQueryOptions) => {
```
Don't need to specify types twice here. Similar comments for other methods in this class.

```suggestion
  transformPage = (wdsPage: RecordQueryResponse, recordType: string, queryOptions: EntityQueryOptions): EntityQueryResponse => {
```

nawatts commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -689,6 +715,7 @@ const WorkspaceData = _.flow(
                       forceRefresh()
                     },
                     after: h(DataTableActions, {
+
```
Formatting nit: this line could be removed.

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,117 @@
+}
+
+
+export class WDSDataTableProvider implements DataTableProvider {
```
I highly recommend that a WdsDataTableProvider interface be defined that extends DataTableProvider and lists the additional public method signatures, etc.  like transformPage method below.

This greatly assists with mocking classes like this when they are the dependency of something else being unit tested.

Also, nit: on the capitalization of WDSDataTableProvider here.... Achrnonyms longer then 2 letters should go cammel-case:  WdsDataTableProvider.

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+
+const uuid = '123e4567-e89b-12d3-a456-426614174000' // value doesn't matter for these tests
+
+const provider = new WDSDataTableProvider(uuid)
```
a provider should be instantiated inside each unit test to avoid potential cross-test state creep in the class under test.

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+describe('WDSDataTableProvider', () => {
+  describe('transformPage', () => {
+    it('restructures a WDS response', () => {
+      // example response from WDS, copy-pasted from a WDS swagger call
```
// Arrange
( please follow the AAA pattern for unit testing - Arrange, Act, Assert, with labeled section comments in each test)

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+      }
+
+      // calling transformPage() on the WDS input should produce this result
+      const expected: EntityQueryResponse = {
```
I recommend moving down to top of // Assert section (see comment below)

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+          filteredPageCount: -1
+        }
+      }
+
```
// Act

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+
+      // transformPage() is the method under test
+      const actual: EntityQueryResponse = provider.transformPage(wdsPage, recordType, queryOptions)
+
```
// Assert

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.test.ts:
```diff
@@ -0,0 +1,165 @@
+  })
+
+  describe('transformMetadata', () => {
+    it('restructures a WDS response', () => {
```
same as above test, please add AAA section headers.

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,117 @@
+interface SearchRequest {
+  offset: number,
+  limit: number,
+  sort: 'asc' | 'ASC' | 'desc' | 'DESC',
```
hmmm... stuff like this casing variation makes me want to recommend that things like this get "scrubbed" and normalized in some way in the data/provider layer of the UI codebase.... 

i.e. convert SearchRequestRaw to SearchRequest with the cleaned up sort property so that downstream UI code doesn't have to fuss with the variations

msilva-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -458,7 +468,7 @@ const DataTableFeaturePreviewFeedbackBanner = () => {
-const workspaceDataTypes = Utils.enumify(['entities', 'entitiesVersion', 'snapshot', 'referenceData', 'localVariables', 'bucketObjects'])
+const workspaceDataTypes = Utils.enumify(['entities', 'entitiesVersion', 'snapshot', 'referenceData', 'localVariables', 'bucketObjects', 'wds'])
 
 const WorkspaceData = _.flow(
```
unit tests for the code changes being made here?

msilva-broad commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,117 @@
+    }
+  }
+
+  getPage: GetPageFn = async (signal: AbortSignal, entityType: string, queryOptions: EntityQueryOptions) => {
```
similar to other comments, the GetPageFn type here can be removed, and add return type to signature to the right of = sign.  (same for methods below this)

also, these methods need unit tests (with mocked Ajax.WorkspaceDataService to avoid real network calls)

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
remove extraneous export {}

davidangb committed:
use Record ts class for maps

davidangb committed:
remove empty line

davidangb committed:
moving WDS ajax to own file

davidangb committed:
clean up method signatures in providers

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
cleaner signatures

davidangb committed:
unit test isolation

davidangb committed:
arg names, extraneous semicolon

nawatts commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -544,6 +557,19 @@ const WorkspaceData = _.flow(
+    if (isFeaturePreviewEnabled('workspace-data-service') && !getConfig().isProd) {
+      try {
+        setWdsSchema([])
+        setWdsSchemaError(false)
```
Nit: Maybe null or undefined instead of false? I feel like a type of `ObjectType | null` or `ObjectType | undefined` for this state variable would be more conventional than `ObjectType | false`.

msilva-broad replied:
my vote is null usually, since it feels more purposeful then undefined.

nawatts commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -732,14 +758,44 @@ const WorkspaceData = _.flow(
-              }, ['Coming soon.']),
-              div({}, '')
+              [
+                wdsSchemaError && h(NoDataPlaceholder, {
```
It's really a problem with DataTypeSection vs this code, but I think this will result in an a11y warning since DataTypeSection has a role of list and NoDataPlaceholder is not a list item. I don't think it's necessary to address here.

nawatts commented on src/libs/ajax/data-table-providers/WDSDataTableProvider.ts:
```diff
@@ -0,0 +1,118 @@
+    return this.transformPage(wdsPage, entityType, queryOptions)
+  }
+
+  transformMetadata = (wdsSchema: RecordTypeSchema[]): EntityMetadata => {
```
Does this need to be a method of this class? Could it be a function exported by this module?

nawatts commented on src/libs/ajax/WorkspaceDataService.ts:
```diff
@@ -0,0 +1,25 @@
+import { SearchRequest } from 'src/libs/ajax/data-table-providers/WDSDataTableProvider'
+
+
+export const WorkspaceDataService = signal => ({
```
It would be nice to have return types for these functions.

nawatts commented on src/libs/ajax/WorkspaceDataService.ts:
```diff
@@ -0,0 +1,25 @@
+  downloadTsv: async (instanceId: string, recordType: string) => {
+    const res = await fetchWDS(`${instanceId}/tsv/v0.2/${recordType}`, _.merge(authOpts(), { signal }))
+    return res
+  }
```
To make methods more consistent in returning response data vs the Response object, could this function return the response data as a blob instead of the Response itself?

```suggestion
  downloadTsv: async (instanceId: string, recordType: string): Promise<Blob> => {
    const res = await fetchWDS(`${instanceId}/tsv/v0.2/${recordType}`, _.merge(authOpts(), { signal }))
    const blob = await res.blob()
    return blob
  }
```

nawatts commented on src/libs/ajax/WorkspaceDataService.ts:
```diff
@@ -0,0 +1,25 @@
+  deleteTable: async (instanceId: string, recordType: string) => {
+    const res = await fetchWDS(`${instanceId}/types/v0.2/${recordType}`,
+      _.mergeAll([authOpts(), { signal, method: 'DELETE' }]))
+    return res
```
To make methods more consistent about returning response data vs the Response object, does this need to return the Response?

```suggestion
  deleteTable: async (instanceId: string, recordType: string): Promise<void> => {
    await fetchWDS(`${instanceId}/types/v0.2/${recordType}`,
      _.mergeAll([authOpts(), { signal, method: 'DELETE' }]))
```

nawatts commented on src/libs/ajax/data-table-providers/DataTableProvider.ts:
```diff
@@ -9,9 +9,9 @@ export interface EntityMetadata {
 }
 
-export type EntityQuerySortDirection = 'asc' | 'desc'
+export type EntityQuerySortDirection = 'asc' | 'ASC' | 'desc' | 'DESC'
```
Would it be preferable to define this type using a consistent case for consumers and have the provider translate internally if needed?

davidangb committed:
explicit code comments for arrange/act/assert

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
undefined instead of false

davidangb committed:
WDSDataTableProvider -> WdsDataTableProvider

davidangb committed:
remove unused getMetadata function

davidangb committed:
move transformMetdata to module export, not provider-class export

davidangb committed:
Delete WDSDataTableProvider.test.ts

davidangb committed:
Delete WDSDataTableProvider.ts

davidangb committed:
normalize capitalization

davidangb committed:
restructure unit test to reflect WdsDataTableProvider structure

davidangb committed:
add return types for WorkspaceDataService ajax methods

davidangb committed:
return blob from downloadTsv

davidangb committed:
unit tests for WdsDataTableProvider

davidangb commented:
@msilva-broad @nawatts I believe I have addressed all your comments (finally), with a few changes/comments:
* I have not extracted a new `WdsDataTableProvider` interface. Instead, I have made one of the offending methods protected, and I have moved the other out of the `WdsDataTableProvider` class and into its containing module.
* I have doubled down on the `deleteTable` method returning a `Response`; I would like to add the response code to the mixpanel event in a future PR
* I did add unit tests for all the ajax-related methods in `WdsDataTableProvider`. I am not convinced of their value. They seem to be testing whether or not I did a good job writing mocks for the back end, but not much else. I won't die on this hill.
* the comment threads on this PR are long and my apologies if I missed anything

Please re-review when you can!

nawatts commented on src/libs/ajax/data-table-providers/DataTableProvider.ts:
```diff
@@ -11,7 +11,7 @@ export interface EntityMetadata {
 export type EntityQuerySortDirection = 'asc' | 'desc'
 
-export type EntityQueryFilterOperator = 'and' | 'or'
+export type EntityQueryFilterOperator = 'and'| 'or'
```
Formatting nit: spacing.

nawatts commented on src/libs/ajax/data-table-providers/DataTableProvider.ts:
```diff
@@ -55,10 +55,13 @@ export type GetPageFn = (signal: AbortSignal, entityType: string, queryOptions:
 export type GetMetadataFn = (signal: AbortSignal) => Promise<EntityMetadata>
 
-export type DeleteTableFn = (entityType: string) => Promise<void>
+export type DeleteTableFn = (entityType: string) => Promise<Response>
```
I'd still vote for this function to extract what it needs from the response (the HTTP status code was missing) and return that vs returning entire response. But I won't request changes over it.

msilva-broad replied:
I think I feel more strongly about this.  We should not be burdening the rest of the UI code to know about dealing with fetch response properties/mechanics.

msilva-broad replied:
Looking at the error handling mechanics in place, I think it's a bigger lift to get us to a point where shielding the rest of the UI code from consuming html error status codes is more straightforward.  So I'm ok with deferring this concern for now.

nawatts commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.ts:
```diff
@@ -0,0 +1,113 @@
+    supportsFiltering: false
+  }
+
+  protected transformPage = (wdsPage: RecordQueryResponse, recordType: string, queryOptions: EntityQueryOptions): EntityQueryResponse => {
```
This and `wdsToEntityServiceMetadata` appear to both transform a WDS response into a Rawls response. Would it be preferable to have them both implemented similarly (either both functions at the module level or both protected methods)?

nawatts replied:
Never mind, I see that these differ in that `wdsToEntityServiceMetadata` is called by another module.

nawatts commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.ts:
```diff
@@ -0,0 +1,113 @@
+  }
+
+  getPage = async (signal: AbortSignal, entityType: string, queryOptions: EntityQueryOptions): Promise<EntityQueryResponse> => {
+    const wdsPage: RecordQueryResponse = await Ajax(signal).WorkspaceDataService
```
Now that WorkspaceDataService methods have return types, TS can infer the type of this variable.

davidangb committed:
whitespace fix

msilva-broad commented on src/libs/ajax.js:
```diff
@@ -1249,7 +1250,8 @@ export const Ajax = signal => {
     OAuth2: OAuth2(signal),
-    Surveys: Surveys(signal)
+    Surveys: Surveys(signal),
+    WorkspaceDataService: WorkspaceDataService(signal)
```
naming on this is not consistent with the other ajax sub-areas. 

I Suggest rename to WorkspaceData (on prop and the export being pulled in here)

msilva-broad commented on src/pages/workspaces/workspace/Data.js:
```diff
@@ -969,6 +1025,13 @@ const WorkspaceData = _.flow(
+            workspace,
+            recordType: selectedData.entityType,
+            wdsSchema
           })]
```
consideration for future:
this is a pretty big file.  I would recommend starting to split this off into sub-components.... and aligning the data-call needs with useInterestingData hook pattern that I paired with Nick on and will be promoting for broad usage.  That will make unit testing this code way easier (which we are starting to push on harder and harder for PRs)

See: GCSFileBrowserProvider

msilva-broad commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.test.ts:
```diff
@@ -0,0 +1,302 @@
+
+  describe('transformPage', () => {
+    it('restructures a WDS response', () => {
+      // ====== Arrange
```
The agreed standard on these headers is 
// Arrange
// Act
// Assert

At other companies, I've gone with ALL CAPS version of that, but above is where we landed here.  A blank line between the sections helps break it up too.

msilva-broad commented on src/libs/ajax/data-table-providers/WdsDataTableProvider.test.ts:
```diff
@@ -0,0 +1,302 @@
+
+
+describe('WdsDataTableProvider', () => {
+  let getRecords
```
some added type safety would be great here... try 
`let getRecords: ReturnType<typeof WorkspaceDataService>['getRecords'] = ...function block and body here...`

and similar for the other methods being mocked here.

```
This should let you then set the props of WorkspaceDataService in the ajax mock further down as:
{
   getRecords: jest.fn().mockImplementation(getRecords),
   ...
}
```

One or two other patterns with good type safety are floating around as well... the functional instantiation of Ajax and the challenges in mocking it creates is a known issue.  But it may take a while before we can consider a pattern rework.

msilva-broad replied:
exporting a WorkspaceDataServiceContract type from where WorkspaceDataService is defined would be helpful too, and shorten the type annotations here.  (But see other comment about renaming to WorkspaceData)

msilva-broad replied:
an easier alternative would be to just add type decorations to these three vars:

`let getRecords: jest.MockedFunction<ReturnType<typeof WorkspaceData>['getRecords']>`

and similar for the other two.

My concern is that the soft cast near the end is not providing enough safety and is not alerting of type drift as well and early as this suggested change would.

`... as ReturnType<typeof Ajax>`

This should be a fairly easy drop-in... but if not, I won't block the PR for it.

davidangb replied:
adding the type decorations to the `let` vars causes type-safety problems on the Ajax mock. I can make the Ajax mock work if I first cast it to unknown:
```typescript
asMockedFn(Ajax).mockImplementation(() => ({ WorkspaceData: { getRecords, deleteTable, downloadTsv } } as unknown as ReturnType<typeof Ajax>))
```
is that worth it?

msilva-broad replied:
this is likely because you are only providing a small set of the total ajax contract.  First, for convenience, define 

`type AjaxContract = ReturnType<typeof Ajax>` 

Then casting the above to `as Partial<AjaxContract> as AjaxContract`

davidangb committed:
rename WorkspaceDataService -> WorkspaceData

davidangb committed:
proper code comments

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

davidangb committed:
more type safety for ajax mocks

davidangb committed:
more type safety

davidangb committed:
Merge branch 'dev' into da_AJ-585_detectWds

msilva-broad reviewed (approved):
great work!

davidangb committed:
lint fixes post-rebase
